docs: working with stages documentation (#2608)
akuity/kargo: docs/docs/30-how-to-guides/14-working-with-stages.md

subfields instead of being described only by the name of a `Warehouse`.
:::

For each `Stage`, the Kargo controller will periodically check for `Freight`
resources that are newly available for promotion to that `Stage`.

When a `Stage` accepts `Freight` directly from its origin, _all_ new `Freight`
created by that origin (e.g. a `Warehouse` ) are immediately available for
promotion to that `Stage`.

When a `Stage` accepts `Freight` from one or more "upstream" `Stage` resources,
`Freight` is considered available for promotion to that `Stage` only after being
_verified_ in at least one of the upstream `Stage`s. Alternatively, users with
adequate permissions may manually _approve_ `Freight` for promotion to any given
`Stage` without requiring upstream verification.

:::tip
Explicit approvals are a useful method for applying the occasional "hotfix"
without waiting for a `Freight` resource to traverse the entirety of a pipeline.
:::

In the following example, the `test` `Stage` requests `Freight` that has
originated from the `my-warehouse` `Warehouse` and indicates that it will accept
new `Freight` _directly_ from that origin:

```yaml

> apiVersion: kargo.akuity.io/v1alpha1
> kind: Stage
> metadata:
> name: test
> namespace: kargo-demo
> spec:
> requestedFreight:
> - origin:
> kind: Warehouse
> name: my-warehouse
> sources:
> direct: true

Reviewer: What happened to the formatting?

```suggestion
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: my-warehouse
    sources:
      direct: true
  # ...
```

# ...
```

In this example, the `uat` `Stage` requests `Freight` that has originated from
the `my-warehouse` `Warehouse`, but indicates that it will accept such `Freight`
only after it has been _verified_ in the `test` `Stage`:

```yaml

> apiVersion: kargo.akuity.io/v1alpha1
> kind: Stage
> metadata:
> name: uat
> namespace: kargo-demo
> spec:
> requestedFreight:
> - origin:
> kind: Warehouse
> name: my-warehouse
> sources:
> stages:
> - test
> # ...

Reviewer: ```suggestion
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: uat
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: my-warehouse
    sources:
      stages:
      - test
  # ...
```

```

Stages may also request `Freight` from multiple sources. The following example
illustrates a `Stage` that requests `Freight` from both a `microservice-a` and
`microservice-b` `Warehouse`:

```yaml

> apiVersion: kargo.akuity.io/v1alpha1
> kind: Stage
> metadata:
> name: test
> namespace: kargo-demo
> spec:
> requestedFreight:
> - origin:
> kind: Warehouse
> name: microservice-a
> sources:
> direct: true
> - origin:
> kind: Warehouse
> name: microservice-b
> sources:
> direct: true
> # ...

Reviewer: ```suggestion
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: microservice-a
    sources:
      direct: true
  - origin:
      kind: Warehouse
      name: microservice-b
    sources:
      direct: true
  # ...
```

the delivery of microservices that are developed and deployed in parallel,
although other uses of this feature are anticipated in the future.
:::

### Promotion Templates

The `spec.promotionTemplate` field is used to describe _how_ to transition
`Freight` into the `Stage`. The `spec.promotionTemplate.steps` field describes
the discrete steps of a promotion process in detail.

In the following, very common example, the `promotionTemplate` describes steps
to:

1. Clone a Git repository containing Kubernetes manifests and Kustomize
configuration, checking out two different branches to two different
directories.

1. Clears the contents of one working tree, with intentions to fully replace its
contents.

1. Runs the equivalent of `kustomize edit set image` to update a
`kustomization.yaml` file with a reference to an updated
`public.ecr.aws/nginx/nginx` container image.

1. Renders the updated manifests using the equivalent of `kustomize build`.

1. Commits the updated manifests and pushes them to the `stage/test` of the
remote repository.

1. Forces Argo CD to sync the `kargo-demo-test` application to the latest commit
of the `stage/test` branch.

```yaml

> promotionTemplate:
> spec:
> steps:
> - uses: git-clone
> config:
> repoURL: https://github.com/example/repo.git
> checkout:
> - fromFreight: true
> path: ./src
> - branch: stage/test
> create: true
> path: ./out
> - uses: git-clear
> config:
> path: ./out
> - uses: kustomize-set-image
> as: update-image
> config:
> path: ./src/base
> images:
> - image: public.ecr.aws/nginx/nginx
> - uses: kustomize-build
> config:
> path: ./src/stages/test
> outPath: ./out
> - uses: git-commit
> as: commit
> config:
> path: ./out
> messageFromSteps:
> - update-image
> - uses: git-push
> config:
> path: ./out
> targetBranch: stage/test
> - uses: argocd-update
> config:
> apps:
> - name: kargo-demo-test
> sources:
> - repoURL: https://github.com/example/repo.git
> desiredCommitFromStep: commit

Reviewer: ```suggestion
promotionTemplate:
  spec:
    steps:
    - uses: git-clone
      config:
        repoURL: https://github.com/example/repo.git
        checkout:
        - fromFreight: true
          path: ./src
        - branch: stage/test
          create: true
          path: ./out
    - uses: git-clear
      config:
        path: ./out
    - uses: kustomize-set-image
      as: update-image
      config:
        path: ./src/base
        images:
        - image: public.ecr.aws/nginx/nginx
    - uses: kustomize-build
      config:
        path: ./src/stages/test
        outPath: ./out
    - uses: git-commit
      as: commit
      config:
        path: ./out
        messageFromSteps:
        - update-image
    - uses: git-push
      config:
        path: ./out
        targetBranch: stage/test
    - uses: argocd-update
      config:
        apps:
        - name: kargo-demo-test
          sources:
          - repoURL: https://github.com/example/repo.git
            desiredCommitFromStep: commit
```

```

__For complete documentation of all of Kargo's built-in promotion steps, refer
to the [Promotion Steps Reference](../35-references/10-promotion-steps.md).__

### Verifications

The `spec.verification` field is used to describe optional verification
processes that should be executed after a `Promotion` has successfully deployed
`Freight` to a `Stage`, and if applicable, after the `Stage` has reached a
healthy state.

Verification processes are defined through _references_ to one or more
[Argo Rollouts `AnalysisTemplate` resources](https://argoproj.github.io/argo-rollouts/features/analysis/)
that reside in the same `Project`/`Namespace` as the `Stage` resource.

:::info
Argo Rollouts `AnalysisTemplate` resources (and the `AnalysisRun` resources that
are spawned from them) were intentionally built to be re-usable in contexts
other than Argo Rollouts. Re-using this resource type to define verification
processes means those processes benefit from this rich and battle-tested feature
of Argo Rollouts.
:::

The following example depicts a `Stage` resource that references an
`AnalysisTemplate` named `kargo-demo` to validate the `test` `Stage` after any
successful `Promotion`:

```yaml

> apiVersion: kargo.akuity.io/v1alpha1
> kind: Stage
> metadata:
> name: test
> namespace: kargo-demo
> spec:
> # ...
> verification:
> analysisTemplates:
> - name: kargo-demo

Reviewer: ```suggestion
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
```

```

It is also possible to specify additional labels, annotations, and arguments
that should be applied to `AnalysisRun` resources spawned from the referenced
`AnalysisTemplate`:

```yaml

> apiVersion: kargo.akuity.io/v1alpha1
> kind: Stage
> metadata:
> name: test
> namespace: kargo-demo
> spec:
> # ...
> verification:
> analysisTemplates:
> - name: kargo-demo
> analysisRunMetadata:
> labels:
> foo: bar
> annotations:
> bat: baz
> args:
> - name: foo
> value: bar

Reviewer: ```suggestion
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
    analysisRunMetadata:
      labels:
        foo: bar
      annotations:
        bat: baz
    args:
    - name: foo
      value: bar
```

```

An `AnalysisTemplate` could be as simple as the following, which merely executes
a Kubernetes `Job` that is defined inline:

```yaml

> apiVersion: argoproj.io/v1alpha1
> kind: AnalysisTemplate
> metadata:
> name: kargo-demo
> namespace: kargo-demo
> spec:
> metrics:
> - name: test
> provider:
> job:
> metadata:
> spec:
> backoffLimit: 1
> template:
> spec:
> containers:
> - name: test
> image: alpine:latest
> command:
> - sleep
> - "10"
> restartPolicy: Never

Reviewer: ```suggestion
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  metrics:
  - name: test
    provider:
      job:
        metadata:
        spec:
          backoffLimit: 1
          template:
            spec:
              containers:
              - name: test
                image: alpine:latest
                command:
                - sleep
                - "10"
              restartPolicy: Never
```

```

:::note
Please consult the
[relevant sections](https://argoproj.github.io/argo-rollouts/features/analysis/)
of the Argo Rollouts documentation for comprehensive coverage of the full range
of `AnalysisTemplate` capabilities.
:::

To manage this verification process for `Stage`s you can use the `kargo verify` command. It allows you to rerun or abort the verification of a `Stage`'s current `Freight`.

To rerun the verification for `test` `Stage`:
```shell
kargo verify stage --project=kargo-demo test
```

If you need to stop an ongoing verification process in a `Stage`, you can use the `--abort` flag:
```shell
kargo verify stage --project=kargo-demo test --abort

> ```

Reviewer: This still feels misplaced. Everything up to this point in the page has discussed the structure of the `Stage` type and now, all of the sudden, we're jumping into CLI usage. It doesn't fit.

See [this comment](https://github.com/akuity/kargo/pull/2608/files#r1844184704) about improving how the page is organized.

finishedAt: "2024-07-15T22:13:25Z"
freight:
commits:
- healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
id: 961cfaedbc53aacdb65110028839a2c1c281290d
repoURL: https://github.com/example/kargo-demo.git
name: 666209fd9755a1e48bec6b27f5f447747410dd9e
origin:
kind: Warehouse
name: kargo-demo
freightCollection:
id: 101bca5b0e18ca7913978a1da956308d2544f741
items:
Warehouse/kargo-demo:
commits:
- healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
id: 961cfaedbc53aacdb65110028839a2c1c281290d
repoURL: https://github.com/example/kargo-demo.git
name: 666209fd9755a1e48bec6b27f5f447747410dd9e
origin:
kind: Warehouse
name: kargo-demo
verificationHistory:
- analysisRun:
name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5
namespace: kargo-demo-09
phase: ""
id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
phase: Pending
startTime: "2024-07-15T22:13:34Z"
phase: Succeeded
observedGeneration: 1
phase: Steady
```

## Refresh

Refreshing a `Stage` helps update its state based on any new changes in `Freight` or configuration. By refreshing, you can ensure that the `Stage` reflects the latest deployment status or artifact updates.

```shell
kargo refresh stage --project=kargo-demo staging
```

> Running this command updates the `staging` `Stage` of the `kargo-demo` project to reflect the most recent `Freight` or configuration changes.

Reviewer: It's worth documenting, but as noted in other comments, let's pull verification details down to here as well and also add docs for other important Stage-related commands.

---
description: Learn how to work effectively with stages

> sidebar_label: Working with stages

Reviewer: ```suggestion
sidebar_label: Working with Stages
```

---

# Working With Stages

> Each Kargo stage is represented by a Kubernetes resource of type `Stage`.

Reviewer: ```suggestion
Each Kargo Stage is represented by a Kubernetes resource of type `Stage`.
```

> 

Reviewer: ```suggestion

## The `Stage` Resource Type

```

> A `Stage` resource's `spec` field decomposes into three main areas of concern:

Reviewer: ```suggestion
Like most Kubernetes resources, a `Stage` is composed of a user-defined `spec` field and a system-populated `status` field.

A `Stage` resource's `spec` field is itself composed of three main areas of concern:
```

* Requested Freight

* Promotion template

* Verification

> The following sections will explore each of these in greater detail.

Reviewer: ```suggestion
The following sections will explore each these `spec` sections as well as `status` in greater detail.
```

> ### Requested Freight

Reviewer: ```suggestion
## Requested Freight
```

kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: my-warehouse
    sources:
      direct: true
  # ...
# ...
```

In this example, the `uat` `Stage` requests `Freight` that has originated from
the `my-warehouse` `Warehouse`, but indicates that it will accept such `Freight`
only after it has been _verified_ in the `test` `Stage`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: uat
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: my-warehouse
    sources:
      stages:
      - test
  # ...
```

Stages may also request `Freight` from multiple sources. The following example
illustrates a `Stage` that requests `Freight` from both a `microservice-a` and
`microservice-b` `Warehouse`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: microservice-a
    sources:
      direct: true
  - origin:
      kind: Warehouse
      name: microservice-b
    sources:
      direct: true
  # ...
```

:::tip

> By requesting `Freight` from multiple sources, a `Stage` can effectively
> participate in _multiple pipelines_ that may each deliver different collections
> of artifacts independently of the others. At present, this is most useful for
> the delivery of microservices that are developed and deployed in parallel,
> although other uses of this feature are anticipated in the future.

Reviewer: ```suggestion
By requesting `Freight` from multiple sources, a `Stage` can effectively
participate in _multiple pipelines_ that may each deliver different collections
of artifacts independently of the others.
```

Removing this because we're already seeing this used in a number of different ways.

:::

> ### Promotion Templates

Reviewer: ```suggestion
## Promotion Templates
```

of the `stage/test` branch.

```yaml
promotionTemplate:
  spec:
    steps:
    - uses: git-clone
      config:
        repoURL: https://github.com/example/repo.git
        checkout:
        - fromFreight: true
          path: ./src
        - branch: stage/test
          create: true
          path: ./out
    - uses: git-clear
      config:
        path: ./out
    - uses: kustomize-set-image
      as: update-image
      config:
        path: ./src/base
        images:
        - image: public.ecr.aws/nginx/nginx
    - uses: kustomize-build
      config:
        path: ./src/stages/test
        outPath: ./out
    - uses: git-commit
      as: commit
      config:
        path: ./out
        messageFromSteps:
        - update-image
    - uses: git-push
      config:
        path: ./out
        targetBranch: stage/test
    - uses: argocd-update
      config:
        apps:
        - name: kargo-demo-test
          sources:
          - repoURL: https://github.com/example/repo.git
            desiredCommitFromStep: commit
```

__For complete documentation of all of Kargo's built-in promotion steps, refer
to the [Promotion Steps Reference](../35-references/10-promotion-steps.md).__

> ### Verifications

Reviewer: ```suggestion
## Verifications
```

      freight:
        commits:
        - healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
          id: 961cfaedbc53aacdb65110028839a2c1c281290d
          repoURL: https://github.com/example/kargo-demo.git
        name: 666209fd9755a1e48bec6b27f5f447747410dd9e
        origin:
          kind: Warehouse
          name: kargo-demo
      freightCollection:
        id: 101bca5b0e18ca7913978a1da956308d2544f741
        items:
          Warehouse/kargo-demo:
            commits:
            - healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
              id: 961cfaedbc53aacdb65110028839a2c1c281290d
              repoURL: https://github.com/example/kargo-demo.git
            name: 666209fd9755a1e48bec6b27f5f447747410dd9e
            origin:
              kind: Warehouse
              name: kargo-demo
        verificationHistory:
        - analysisRun:
            name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5
            namespace: kargo-demo-09
            phase: ""
          id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
          phase: Pending
          startTime: "2024-07-15T22:13:34Z"
      phase: Succeeded
  observedGeneration: 1
  phase: Steady
```

> 

Reviewer: Right here seems to be the right place to add something like `## Interacting with Stages`.

You can show how to create, delete, and refresh Stages. Info on promoting _to_ a Stage probably belongs in a not-yet-written "Working with Promotions" page, so I would not include it here (but you can link to it later when it exists). Verifying the Stage's current Freight could also be discussed here. i.e. Refer back to [this comment](https://github.com/akuity/kargo/pull/2608/files#r1844211810) and relocate those details to this new section.

For each of create, delete, refresh, and verify, you can show how these can be accomplished with either the UI or CLI.

---

> description: Learn how to work effectively with stages

Reviewer: ```suggestion
description: Learn how to work effectively with Stages
```

In the following, very common example, the `promotionTemplate` describes steps
to:

1. Clone a Git repository containing Kubernetes manifests and Kustomize
   configuration, checking out two different branches to two different
   directories.

1. Clears the contents of one working tree, with intentions to fully replace its
   contents.

1. Runs the equivalent of `kustomize edit set image` to update a
   `kustomization.yaml` file with a reference to an updated
   `public.ecr.aws/nginx/nginx` container image.

1. Renders the updated manifests using the equivalent of `kustomize build`.

1. Commits the updated manifests and pushes them to the `stage/test` of the
   remote repository.

1. Forces Argo CD to sync the `kargo-demo-test` application to the latest commit
   of the `stage/test` branch.

```yaml
promotionTemplate:
  spec:
    vars:
    - name: gitopsRepo
      value: https://github.com/example/repo.git
    - name: imageRepo
      value: public.ecr.aws/nginx/nginx
    - name: srcPath
      value: ./src
    - name: outPath
      value: ./out
    - name: targetBranch
      value: stage/${{ ctx.stage }}
    steps:
    - uses: git-clone
      config:

>         repoURL: \${{ vars.gitopsRepo }}
>         checkout:
>         - branch: main
>           path: \${{ vars.srcPath }}
>         - branch: stage/\${{ ctx.stage }}
>           create: true
>           path: \${{ vars.outPath }}
>     - uses: git-clear
>       config:
>         path: \${{ vars.outPath }}
>     - uses: kustomize-set-image
>       as: update-image
>       config:
>         path: \${{ vars.srcPath }}/base
>         images:
>         - image: \${{ vars.imageRepo }}
>     - uses: kustomize-build
>       config:
>         path: \${{ vars.srcPath }}/stages/\${{ ctx.stage }}
>         outPath: \${{ vars.outPath }}/manifests.yaml
>     - uses: git-commit
>       as: commit
>       config:
>         path: \${{ vars.outPath }}
>         messageFromSteps:
>         - update-image
>     - uses: git-push
>       config:
>         path: \${{ vars.outPath }}
>         branch: \${{ vars.targetBranch }}
>     - uses: argocd-update
>       config:
>         apps:
>         - name: kargo-demo-\${{ ctx.stage }}
>           sources:
>           - repoURL: \${{ vars.gitopsRepo }}
>             desiredRevision: \${{ outputs.commit.commit }}

Reviewer: The escaping of the `$` character isn't necessary here. If you've seen it in other places, they are places where there was YAML being pasted into a terminal and the shell had to be made to ignore the `$`.

```suggestion
        repoURL: ${{ vars.gitopsRepo }}
        checkout:
        - branch: main
          path: ${{ vars.srcPath }}
        - branch: stage/${{ ctx.stage }}
          create: true
          path: ${{ vars.outPath }}
    - uses: git-clear
      config:
        path: ${{ vars.outPath }}
    - uses: kustomize-set-image
      as: update-image
      config:
        path: ${{ vars.srcPath }}/base
        images:
        - image: ${{ vars.imageRepo }}
    - uses: kustomize-build
      config:
        path: ${{ vars.srcPath }}/stages/${{ ctx.stage }}
        outPath: ${{ vars.outPath }}/manifests.yaml
    - uses: git-commit
      as: commit
      config:
        path: ${{ vars.outPath }}
        messageFromSteps:
        - update-image
    - uses: git-push
      config:
        path: ${{ vars.outPath }}
        branch: ${{ vars.targetBranch }}
    - uses: argocd-update
      config:
        apps:
        - name: kargo-demo-${{ ctx.stage }}
          sources:
          - repoURL: ${{ vars.gitopsRepo }}
            desiredRevision: ${{ outputs.commit.commit }}
```

spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
    analysisRunMetadata:
      labels:
        foo: bar
      annotations:
        bat: baz
    args:
    - name: foo
      value: bar
```

An `AnalysisTemplate` could be as simple as the following, which merely executes
a Kubernetes `Job` that is defined inline:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  metrics:
  - name: test
    provider:
      job:
        metadata:
        spec:
          backoffLimit: 1
          template:
            spec:
              containers:
              - name: test
                image: alpine:latest
                command:
                - sleep
                - "10"
              restartPolicy: Never
```

:::note
Please consult the
[relevant sections](https://argoproj.github.io/argo-rollouts/features/analysis/)
of the Argo Rollouts documentation for comprehensive coverage of the full range
of `AnalysisTemplate` capabilities.
:::

### Status

The `status` field of a `Stage` resource records:

  * Conditions containing the last observations of the `Stage`'s current state.

>   * The current phase of the `Stage`'s lifecycle.

Reviewer: ```suggestion
  * The current phase of the `Stage`'s lifecycle (distilled from the conditions).
```

  * Details about the last `Promotion` and any in-progress `Promotion`.
  * History of `Freight` that has been deployed to the `Stage` (from most to
  least recent) along with the results of any associated verification processes.
  * The health status of related Argo CD `Application` resources.

For example:

```yaml
status:
  freightHistory:
  - id: 101bca5b0e18ca7913978a1da956308d2544f741
    items:
      Warehouse/my-warehouse:
        commits:
        - healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
          id: 961cfaedbc53aacdb65110028839a2c1c281290d

>           repoURL: https://github.com/example/kargo-demo.git

Reviewer: I want to make this look more generic so readers focus on the structure of the status instead of getting absorbed in the details of what might be a real example. `https://github.com/example/repo.git` is easily understood to be a fake repo.

```suggestion
          repoURL: https://github.com/example/repo.git
```

        images:
        - digest: sha256:b2487a28589657b318e0d63110056e11564e73b9fd3ec4c4afba5542f9d07d46
          repoURL: public.ecr.aws/nginx/nginx
          tag: 1.27.0
        name: 666209fd9755a1e48bec6b27f5f447747410dd9e
        origin:
          kind: Warehouse
          name: my-warehouse
    verificationHistory:
    - analysisRun:
        name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5

>         namespace: kargo-demo-09

Reviewer: The `09` feels distracting. I know it was there to begin with, but let's kill it now.

```suggestion
        namespace: kargo-demo
```

        phase: Successful
      finishTime: "2024-07-15T22:13:57Z"
      id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
      phase: Successful
      startTime: "2024-07-15T22:13:34Z"
  health:
    argoCDApps:
    - healthStatus:
        status: Healthy

>       name: kargo-demo-09-test

Reviewer: ```suggestion
      name: kargo-demo-test
```

      name: 666209fd9755a1e48bec6b27f5f447747410dd9e
      origin:
        kind: Warehouse
        name: kargo-demo
    name: test.01j2w7a15cxjjgejresfyw6ysp.666209f
    status:
      finishedAt: "2024-07-15T22:13:25Z"
      freight:
        commits:
        - healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
          id: 961cfaedbc53aacdb65110028839a2c1c281290d
          repoURL: https://github.com/example/kargo-demo.git
        name: 666209fd9755a1e48bec6b27f5f447747410dd9e
        origin:
          kind: Warehouse
          name: kargo-demo
      freightCollection:
        id: 101bca5b0e18ca7913978a1da956308d2544f741
        items:
          Warehouse/kargo-demo:
            commits:
            - healthCheckCommit: 111eaf55aa41f21bb9bb707ba1baa748b83ec51e
              id: 961cfaedbc53aacdb65110028839a2c1c281290d
              repoURL: https://github.com/example/kargo-demo.git
            name: 666209fd9755a1e48bec6b27f5f447747410dd9e
            origin:
              kind: Warehouse
              name: kargo-demo
        verificationHistory:
        - analysisRun:
            name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5

>             namespace: kargo-demo-09

Reviewer: ```suggestion
            namespace: kargo-09
```

            phase: ""
          id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
          phase: Pending
          startTime: "2024-07-15T22:13:34Z"
      phase: Succeeded
  observedGeneration: 1
  phase: Steady
```

## Interacting with Stages

> Kargo provides tools to manage Stages using either the Kargo Dashboard or the
> Kargo CLI. This section explains how to handle Stages effectively through both interfaces.

Reviewer: There are a lot of places throughout the docs where we say UI or dashboard. I think we should standardize on one term. We'll have to clean it up in other places as well, but let's get started here.

```suggestion
Kargo provides tools to manage Stages using either its UI or
CLI. This section explains how to handle Stages effectively through both interfaces.

:::info
Users with credentials for and sufficient permissions within the Kargo control plane's Kubernetes cluster can also manage `Stage` resources using `kubectl`.
:::
```

### Creating a Stage

<Tabs groupId="create-stage">

> <TabItem value="ui" label="Using the Kargo Dashboard" default>

Reviewer: ```suggestion
<TabItem value="ui" label="Using the UI" default>
```

> 1. Navigate to your Project dashboard and locate the action menu in the right corner of the pipeline:

Reviewer: ```suggestion
1. Navigate to your Project in the Kargo UI and locate the action menu in the right corner of the pipeline:
```

   ![create-stage](../../static/img/create-stage.png)

> 2. Click the magic wand icon to open the dropdown, then select <Hlt>Create Stage</Hlt>.

Reviewer: ```suggestion
1. Click the magic wand icon to open the dropdown, then select <Hlt>Create Stage</Hlt>.
```

>    A form will appear to input details for the new `Stage`:

Reviewer: ```suggestion
   A form will appear to input details for a new `Stage`:
```

   ![create-stage](../../static/img/create-stage-2.png)

> 3. Complete the form with the necessary details and submit it.

Reviewer: ```suggestion
1. Complete the form with the necessary details and submit it.
```

>    The new `Stage` will be added to the pipeline, connected to other

Reviewer: Separate your commentary on what just happened from the actions you want the readers to take.

```suggestion

   The new `Stage` will be added to the pipeline, connected to other
```

   `Stage`s based on your configuration:

   ![create-stage](../../static/img/create-stage-3.png)

> </TabItem>

Reviewer: ```suggestion

</TabItem>
```

> <TabItem value="cli" label="Using the Kargo CLI">

Reviewer: ```suggestion
<TabItem value="cli" label="Using the CLI">
```

1. Define the `Stage` in a YAML file, for example:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:

>   name: <stage-name>

Reviewer: ```suggestion
  name: <stage>
```

>   namespace: <project-name>

Reviewer: ```suggestion
  namespace: <project>
```

spec:
  ### Add your Stage specifications here
```

> 2. Save the file and run:

Reviewer: ```suggestion
1. Save the file and run:
```

```shell

> kargo create -f <stage-filename>

Reviewer: ```suggestion
kargo create -f <filename>
```

```

> 3. Verify the creation by listing the `Stage`:

Reviewer: ```suggestion
1. Verify the creation by listing `Stage`s:
```

```shell

> kargo get stage <stage-name> --project <project-name>

Reviewer: ```suggestion
kargo get stage <stage> --project <project>
```

```

</TabItem>
</Tabs>

### Deleting a Stage

<Tabs groupId="delete-stage">

> <TabItem value="ui" label="Using the Kargo Dashboard" default>

Reviewer: ```suggestion
<TabItem value="ui" label="Using the UI" default>
```

1. Select the `Stage` you want to remove.

> 2. Click <Hlt>Delete</Hlt> in the top-right corner of the pop-up window:

Reviewer: Changing this only because we used "upper right" instead of "top right" somewhere else.

```suggestion
2. Click <Hlt>Delete</Hlt> in the upper right corner of the pop-up window:
```

   ![delete-stage](../../static/img/delete-stage.png)

> </TabItem>

Reviewer: ```suggestion

</TabItem>
```

<TabItem value="cli" label="Using the Kargo CLI">

> To delete a `Stage` using the CLI, run:

Reviewer: ```suggestion

To delete a `Stage` using the CLI, run:
```

> <TabItem value="cli" label="Using the Kargo CLI">
> To delete a `Stage` using the CLI, run:

Reviewer: ```suggestion
<TabItem value="cli" label="Using the CLI">

To delete a `Stage` using the CLI, run:
```

```shell

> kargo delete stage --project <project-name> <stage-name>

Reviewer: I notice we previously used the `--project` flag _after_ the argument and here we were using it before. We should be consistent.

```suggestion
kargo delete stage <stage> --project <project>
```

```

</TabItem>
</Tabs>

### Refreshing a Stage

> Refreshing a `Stage` triggers a reconciliation process that
> performs essential tasks, including checking for completed 
> Promotions, running health checks, verifying the current 
> Freight, and determining if any pending Promotions can proceed. 
> This ensures that the `Stage` reflects the latest state and operational readiness.

Reviewer: ```suggestion
Refreshing a `Stage` triggers its reconciliation process, which
includes checking for any newly-completed Promotions, queueing
up the next, pending Promotion, when applicable, and executing
any applicable health check processes.
```

<Tabs groupId="refresh-stage">
<TabItem value="ui" label="Using the Kargo Dashboard" default>

> 1. Select the `Stage` you want to refresh.

Reviewer: ```suggestion

1. Select the `Stage` you want to refresh.
```

> <TabItem value="ui" label="Using the Kargo Dashboard" default>
> 1. Select the `Stage` you want to refresh.
> 2. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:

Reviewer: ```suggestion
<TabItem value="ui" label="Using the Dashboard" default>

1. Select the `Stage` you want to refresh.
1. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:
```

   ![refresh-stage](../../static/img/refresh-stage.png)

> </TabItem>

Reviewer: ```suggestion

</TabItem>
```

> <TabItem value="cli" label="Using the Kargo CLI">
> To refresh a `Stage`, run:

Reviewer: ```suggestion
<TabItem value="cli" label="Using the CLI">

To refresh a `Stage`, run:
```

```shell

> kargo refresh stage --project=<project-name> <stage-name>

Reviewer: ```suggestion
kargo refresh stage <stage> --project=<project name>
```

```

</TabItem>
</Tabs>

### Reverifying a Stage's Current Freight

> Reverification ensures that the `Freight` associated with a `Stage` is functioning as expected. This can include rerunning or aborting verification.

Reviewer: ```suggestion
Verification processes, which run automatically following each successful Promotion
can be re-run on-demand. This is useful for re-attempting a failed verification
process or just to validate that applications within the `Stage` are performing
as desired.
```

<Tabs groupId="verify-stage">
<TabItem value="ui" label="Using the Kargo Dashboard" default>

> 1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

Reviewer: ```suggestion

1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:
```

> <TabItem value="ui" label="Using the Kargo Dashboard" default>
> 1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

Reviewer: ```suggestion
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:
```

![verify-stage](../../static/img/reverify-freight.png)

> 2. To confirm the `Stage(s)` where the `Freight` has been verified, return

Reviewer: ```suggestion
1. To confirm the `Stage(s)` where the `Freight` has been verified, return
```

to the <Hlt>Freight Timeline</Hlt> and select the `Freight`.
Verified `Stage` names will appear under <Hlt>VERIFIED IN</Hlt>:

![verify-stage](../../static/img/verified-in.png)

> </TabItem>

Reviewer: ```suggestion

</TabItem>
```

> <TabItem value="cli" label="Using the Kargo CLI">

Reviewer: ```suggestion
<TabItem value="cli" label="Using the CLI">
```

1. To rerun verification using the CLI run:

```shell

> kargo verify stage --project=<project-name> <stage-name>

Reviewer: ```suggestion
kargo verify stage <stage> --project=<project> 
```

```

> 2. To stop an ongoing verification process, use:

Reviewer: This is awkward. It's being presented as:

Step 1: Start verification
Step 2: Abort it

Those feel like two different things, rather than elements of the same sequence.

This also differs from the UI-centric instructions above, which didn't talk about aborting a verification, but went from starting one to showing how you can see which Stages a piece of Freight has been verified in.

We should try to keep things consistent and have the UI and CLI tabs cover the same things in the same order.

```shell

> kargo verify stage --project=<project-name> <stage-name> --abort

Reviewer: ```suggestion
kargo verify stage <stage> --project=<project> --abort
```

```

</TabItem>
</Tabs>

> Note: For detailed instructions on promoting a `Stage`, refer to the Working with Promotions page.

Reviewer: Let's avoid easily missed "Note:" when we have `:::note` at our disposal.

```suggestion
:::note
For detailed instructions on promoting a `Stage`, refer to the Working with Promotions page.
:::
```

"Refer to..." should also be a link.

:::

### Promotion Templates

The `spec.promotionTemplate` field is used to describe _how_ to transition
`Freight` into the `Stage`. The `spec.promotionTemplate.steps` field describes
the discrete steps of a promotion process in detail.

In the following, very common example, the `promotionTemplate` describes steps
to:

1. Clone a Git repository containing Kubernetes manifests and Kustomize
configuration, checking out two different branches to two different
directories.

1. Clears the contents of one working tree, with intentions to fully replace its
contents.

1. Runs the equivalent of `kustomize edit set image` to update a
`kustomization.yaml` file with a reference to an updated
`public.ecr.aws/nginx/nginx` container image.

1. Renders the updated manifests using the equivalent of `kustomize build`.

1. Commits the updated manifests and pushes them to the `stage/test` of the
remote repository.

1. Forces Argo CD to sync the `kargo-demo-test` application to the latest commit
of the `stage/test` branch.

```yaml
promotionTemplate:
  spec:
    steps:
    - uses: git-clone
      config:
        repoURL: https://github.com/example/repo.git
        checkout:

>         - fromFreight: true

Reviewer: This example needs to be updated to favor the use of expressions over deprecated fields such as `fromFreight`. Look at the promotion steps reference or my own examples repo for guidance on what a more up-to-date version of this process looks like.

a Kubernetes `Job` that is defined inline:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  metrics:
  - name: test
    provider:
      job:
        metadata:
        spec:
          backoffLimit: 1
          template:
            spec:
              containers:
              - name: test
                image: alpine:latest
                command:
                - sleep
                - "10"
              restartPolicy: Never
```

:::note
Please consult the
[relevant sections](https://argoproj.github.io/argo-rollouts/features/analysis/)
of the Argo Rollouts documentation for comprehensive coverage of the full range
of `AnalysisTemplate` capabilities.
:::

### Status

A `Stage` resource's `status` field records:

* The current phase of the `Stage` resource's lifecycle.

* Information about the last `Promotion` and any in-progress `Promotion`.

* History of `Freight` that has been deployed to the `Stage` (from most to
least recent) along with the results of any associated verification processes.

* The health status of any associated Argo CD `Application` resources.

For example:

> ```yaml

Reviewer: This whole example probably needs to be brought up to date with the live status captured from a more recent Kargo version. The key things missing are conditions (mentioned [here](https://github.com/akuity/kargo/pull/2608/files#r1882371611) and numerous details of `lastPromotion.status` that were added within the last few releases.

Reviewer: I would actually suggest running Kargo from the head of `main` when doing this, as there are more such changes in `lastPromotion.status` coming in 1.2.

Reviewer: This still looks very out of date.

              id: 961cfaedbc53aacdb65110028839a2c1c281290d
              repoURL: https://github.com/example/kargo-demo.git
            name: 666209fd9755a1e48bec6b27f5f447747410dd9e
            origin:
              kind: Warehouse
              name: kargo-demo
        verificationHistory:
        - analysisRun:
            name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5
            namespace: kargo-demo-09
            phase: ""
          id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
          phase: Pending
          startTime: "2024-07-15T22:13:34Z"
      phase: Succeeded
  observedGeneration: 1
  phase: Steady
```

## Interacting with Stages

Kargo provides tools to manage Stages using either the Kargo Dashboard or the
Kargo CLI. This section explains how to handle Stages effectively through both interfaces.

Kargo simplifies managing `Stage`s through the Kargo Dashboard or the Kargo CLI. This section
explains how to effectively perform key actions—like creating, deleting, refreshing Stages and
verifying it's Freight—using both interfaces.

### Creating a Stage

<Tabs groupId="create-stage">
  <TabItem value="ui" label="Using the Kargo Dashboard" default>

>     Navigate to your Project dashboard and locate the action menu in the right corner of the pipeline:

Reviewer: Indentation of all content in this tab also seems to be off. It renders correctly, but the unnecessary indentations can throw other things off if/when the section is amended in the future.

    ![create-stage](../../static/img/create-stage.png)

    Click the magic wand icon to open the dropdown, then select <Hlt>Create Stage</Hlt>.
    A form will appear to input details for the new `Stage`:

    ![create-stage](../../static/img/create-stage-2.png)

    Complete the form with the necessary details and submit it.
    The new `Stage` will be added to the pipeline, connected to other
    `Stage`s based on your configuration:

    ![create-stage](../../static/img/create-stage-3.png)
  </TabItem>

  <TabItem value="cli" label="Using the Kargo CLI">

>     Define the `Stage` in a YAML file, for example:

Reviewer: Enumerate and fix indentation.

    ```yaml
    apiVersion: kargo.akuity.io/v1alpha1
    kind: Stage
    metadata:
      name: <stage-name>
      namespace: <project-name>
    spec:
      ### Add your Stage specifications here
    ```

    Save the file and run:

    ```shell
    kargo create -f <stage-filename>
    ```

>     Example Output:
>     ```shell
>     stage.kargo.akuity.io/<stage-name> created
>     ```

Reviewer: Let's try to show example output only where it adds some value or when there's something in the output worth discussing. In a case like this, the example doesn't add much.

```suggestion
```

    Verify the creation by listing the `Stage`:

    ```shell
    kargo get stage <stage-name> --project <project-name>
    ```

>     Example Output:
>     ```shell
>     NAME    SHARD   CURRENT FREIGHT   HEALTH   PHASE  AGE
>     test                                              70s
>     ```

Reviewer: Same as [here](https://github.com/akuity/kargo/pull/2608/files#r1882388556).

```suggestion
```

  </TabItem>
</Tabs>

### Deleting a Stage

<Tabs groupId="delete-stage">
  <TabItem value="ui" label="Using the Kargo Dashboard" default>

>     Select the `Stage` you want to remove and click <Hlt>Delete</Hlt> in the top-right corner of the pop-up window:

Reviewer: Enumerate and fix indentation.

    ![delete-stage](../../static/img/delete-stage.png)
  </TabItem>

  <TabItem value="cli" label="Using the Kargo CLI">
    To delete a `Stage` using the CLI, run:

    ```shell
    kargo delete stage --project <project-name> <stage-name>
    ```

>     Example Output:
>     ```shell
>       stage.kargo.akuity.io/<stage-name> deleted
>     ```

Reviewer: ```suggestion
```

  </TabItem>
</Tabs>

### Refreshing a Stage

Refreshing ensures the `Stage` reflects the latest updates,

> including changes in `Freight` or configuration.

Reviewer: I'm not sure I'd describe it this way. It forces the controller to reconcile the Stage, which means:

1. Check for recently completed Promotions
2. Cascade reconciliation to any in-progress Promotion
3. Run health checks
4. Verify or reverify the current Freight
5. Decide whether the next pending Promotion (if any) may now proceed
6. If auto-promotion is enabled, look for new Freight and create corresponding Promotion(s)

Reviewer: idk if you actually need to include this level of detail, but my point is only that when you refresh, real work happens. It's not just a trivial update of the status, which is how I read this.

<Tabs groupId="refresh-stage">
  <TabItem value="ui" label="Using the Kargo Dashboard" default>
    Select the `Stage` you want to refresh and click <Hlt>Refresh</Hlt> in the top-right
    corner of the pop-up window:

    ![refresh-stage](../../static/img/refresh-stage.png)
  </TabItem>

  <TabItem value="cli" label="Using the Kargo CLI">
    To refresh a `Stage`, run:

    ```shell
    kargo refresh stage --project=<project-name> <stage-name>
    ```

>     Example Output:
>     ```shell
>     stage 'kargo-demo/<stage-name>' refreshed
>     ```

Reviewer: ```suggestion
```

  </TabItem>
</Tabs>

### Reverifying a Stage's Current Freight

Reverification ensures that the `Freight` associated with a `Stage` is functioning as expected. This can include rerunning or aborting verification.

<Tabs groupId="verify-stage">
  <TabItem value="ui" label="Using the Kargo Dashboard" default>

>     Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

Reviewer: Enumeration and indenting

    ![verify-stage](../../static/img/reverify-freight.png)

    To confirm the `Stage(s)` where the `Freight` has been verified, return
    to the <Hlt>Freight Timeline</Hlt> and select the `Freight`.
    Verified `Stage` names will appear under <Hlt>VERIFIED IN</Hlt>:

    ![verify-stage](../../static/img/verified-in.png)
  </TabItem>

  <TabItem value="cli" label="Using the Kargo CLI">

>     To rerun verification using the CLI run:

Reviewer: Mind the indents

            name: 666209fd9755a1e48bec6b27f5f447747410dd9e
            origin:
              kind: Warehouse
              name: kargo-demo
        verificationHistory:
        - analysisRun:
            name: test.01j2w7aknhf3j7jteyqs72hnbg.101bca5
            namespace: kargo-demo
            phase: ""
          id: 5535a484-bbd0-4f12-8cf4-be2c8e0041c9
          phase: Pending
          startTime: "2024-07-15T22:13:34Z"
      phase: Succeeded
  observedGeneration: 1
  phase: Steady
```

## Interacting with Stages

Kargo provides tools to manage Stages using either its UI or
CLI. This section explains how to handle Stages effectively through both interfaces.

:::info
Users with credentials for and sufficient permissions within the Kargo control plane's Kubernetes cluster can also manage `Stage` resources using `kubectl`.
:::

### Creating a Stage

<Tabs groupId="create-stage">
<TabItem value="ui" label="Using the UI" default>

1. Navigate to your Project in the Kargo UI and locate the action menu in the upper right corner of the pipeline:

   ![create-stage](../../static/img/create-stage.png)

1. Click the magic wand icon to open the dropdown, then select <Hlt>Create Stage</Hlt>.

>    A form will appear to input details for a new `Stage`:

Reviewer: ```suggestion

   A form will appear to input details for a new `Stage`:
```

   ![create-stage](../../static/img/create-stage-2.png)

1. Complete the form with the necessary details and submit it.

   The new `Stage` will be added to the pipeline, connected to other
   `Stage`s based on your configuration:

   ![create-stage](../../static/img/create-stage-3.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

1. Define the `Stage` in a YAML file, for example:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: <stage>
  namespace: <project>
spec:
  ### Add your Stage specifications here
```

1. Save the file and run:

```shell
kargo create -f <filename>
```

1. Verify the creation by listing `Stage`s:

```shell
kargo get stage <stage> --project <project>
```

</TabItem>
</Tabs>

### Deleting a Stage

<Tabs groupId="delete-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to remove.

> 2. Click <Hlt>Delete</Hlt> in the upper right corner of the pop-up window:

Reviewer: ```suggestion
1. Click <Hlt>Delete</Hlt> in the upper right corner of the pop-up window:
```

> 1. Click <Hlt>Delete</Hlt> in the upper right corner of the pop-up window:

Reviewer: ```suggestion

1. Click <Hlt>Delete</Hlt> in the upper right corner of the pop-up window:
```

   ![delete-stage](../../static/img/delete-stage.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

To delete a `Stage` using the CLI, run:

```shell
kargo delete stage <stage> --project <project>
```

</TabItem>
</Tabs>

### Refreshing a Stage

Refreshing a `Stage` triggers its reconciliation process, which
includes checking for any newly-completed Promotions, queueing
up the next, pending Promotion, when applicable, and executing
any applicable health check processes.

<Tabs groupId="refresh-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to refresh.

> 1. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:

Reviewer: ```suggestion

1. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:
```

   ![refresh-stage](../../static/img/refresh-stage.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

To refresh a `Stage`, run:

```shell
kargo refresh stage <stage> --project <project>
```

</TabItem>
</Tabs>

### Reverifying a Stage's Current Freight

Verification processes, which run automatically following each successful Promotion,
can also be re-run on-demand. This is useful for re-attempting a failed verification
process or just to validate that applications within the `Stage` are performing
as desired.

<Tabs groupId="verify-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

![verify-stage](../../static/img/reverify-freight.png)

1. To see the `Stage`s where the `Freight` has been successfully verified, return
to the <Hlt>Freight Timeline</Hlt> and select the `Freight`.
Verified `Stage` names will appear under <Hlt>VERIFIED IN</Hlt>:

![verify-stage](../../static/img/verified-in.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">
1. To rerun verification using the CLI run:

```shell
kargo verify stage <stage> --project <project> 
```

> 2. To stop an ongoing verification process, use:

Reviewer: ```suggestion
1. To stop an ongoing verification process, use:
```

kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
    analysisRunMetadata:
      labels:
        foo: bar
      annotations:
        bat: baz
    args:
    - name: foo
      value: bar
```

An `AnalysisTemplate` could be as simple as the following, which merely executes
a Kubernetes `Job` that is defined inline:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  metrics:
  - name: test
    provider:
      job:
        metadata:
        spec:
          backoffLimit: 1
          template:
            spec:
              containers:
              - name: test
                image: alpine:latest
                command:
                - sleep
                - "10"
              restartPolicy: Never
```

:::note
Please consult the
[relevant sections](https://argoproj.github.io/argo-rollouts/features/analysis/)
of the Argo Rollouts documentation for comprehensive coverage of the full range
of `AnalysisTemplate` capabilities.
:::

### Status

The `status` field of a `Stage` resource records:

>   * Conditions containing the last observations of the `Stage`'s current state.
>   * The current phase of the `Stage`'s lifecycle (distilled from the conditions).
>   * Details about the last `Promotion` and any in-progress `Promotion`.
>   * History of `Freight` that has been deployed to the `Stage` (from most to
>   least recent) along with the results of any associated verification processes.
>   * The health status of related Argo CD `Application` resources.

Reviewer: ```suggestion
  * Conditions containing the last observations of the `Stage`'s current state.

  * The current phase of the `Stage`'s lifecycle (distilled from the conditions).

  * Details about the last `Promotion` and any in-progress `Promotion`.

  * History of `Freight` that has been deployed to the `Stage` (from most to
    least recent) along with the results of any associated verification processes.

  * The health status of related Argo CD `Application` resources.
```

any applicable health check processes.

<Tabs groupId="refresh-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to refresh.

1. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:

   ![refresh-stage](../../static/img/refresh-stage.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

To refresh a `Stage`, run:

```shell
kargo refresh stage <stage> --project <project>
```

</TabItem>
</Tabs>

### Reverifying a Stage's Current Freight

Verification processes, which run automatically following each successful Promotion,
can also be re-run on-demand. This is useful for re-attempting a failed verification
process or just to validate that applications within the `Stage` are performing
as desired.

<Tabs groupId="verify-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

![verify-stage](../../static/img/reverify-freight.png)

1. To see the `Stage`s where the `Freight` has been successfully verified, return
to the <Hlt>Freight Timeline</Hlt> and select the `Freight`.

> Verified `Stage` names will appear under <Hlt>VERIFIED IN</Hlt>:

Reviewer: ```suggestion

Verified `Stage` names will appear under <Hlt>VERIFIED IN</Hlt>:
```

![verify-stage](../../static/img/verified-in.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

> 1. To rerun verification using the CLI run:

Reviewer: ```suggestion

1. To rerun verification using the CLI run:
```

Refreshing a `Stage` triggers its reconciliation process, which
includes checking for any newly-completed Promotions, queueing
up the next, pending Promotion, when applicable, and executing
any applicable health check processes.

<Tabs groupId="refresh-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to refresh.

1. Click <Hlt>Refresh</Hlt> in the top-right corner of the pop-up window:

   ![refresh-stage](../../static/img/refresh-stage.png)

</TabItem>

<TabItem value="cli" label="Using the CLI">

To refresh a `Stage`, run:

```shell
kargo refresh stage <stage> --project <project>
```

</TabItem>
</Tabs>

### Reverifying a Stage's Current Freight

Verification processes, which run automatically following each successful Promotion,
can also be re-run on-demand. This is useful for re-attempting a failed verification
process or just to validate that applications within the `Stage` are performing
as desired.

<Tabs groupId="verify-stage">
<TabItem value="ui" label="Using the UI" default>

1. Select the `Stage` you want to reverify and click <Hlt>Reverify</Hlt> at the top of the menu:

    ![verify-stage](../../static/img/reverify-freight.png)

>     If you want to stop the verification process mid-way, you can click the <Hlt>Abort Verification</Hlt> button next to the Reverify option.

Reviewer: ```suggestion
    :::note
    If you wish to stop the in-progress verification, you can click <Hlt>Abort Verification</Hlt>.
    :::
```

  # ...
# ...
```

In this example, the `uat` `Stage` requests `Freight` that has originated from
the `my-warehouse` `Warehouse`, but indicates that it will accept such `Freight`
only after it has been _verified_ in the `test` `Stage`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: uat
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: my-warehouse
    sources:
      stages:
      - test
  # ...
```

Stages may also request `Freight` from multiple sources. The following example
illustrates a `Stage` that requests `Freight` from both a `microservice-a` and
`microservice-b` `Warehouse`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: microservice-a
    sources:
      direct: true
  - origin:
      kind: Warehouse
      name: microservice-b
    sources:
      direct: true
  # ...
```

:::tip
By requesting `Freight` from multiple sources, a `Stage` can effectively
participate in _multiple pipelines_ that may each deliver different collections
of artifacts independently of the others.

> :::

Reviewer: ```suggestion

**This is _advanced_ configuration. If you're very new to Kargo,
and requesting `Freight` from multiple origins, you probably
didn't mean to.**
:::
```

      value: ./out
    - name: targetBranch
      value: stage/${{ ctx.stage }}
    steps:
    - uses: git-clone
      config:
        repoURL: ${{ vars.gitopsRepo }}
        checkout:
        - branch: main
          path: ${{ vars.srcPath }}
        - branch: stage/${{ ctx.stage }}
          create: true
          path: ${{ vars.outPath }}
    - uses: git-clear
      config:
        path: ${{ vars.outPath }}
    - uses: kustomize-set-image
      as: update-image
      config:
        path: ${{ vars.srcPath }}/base
        images:
        - image: ${{ vars.imageRepo }}
    - uses: kustomize-build
      config:
        path: ${{ vars.srcPath }}/stages/${{ ctx.stage }}
        outPath: ${{ vars.outPath }}/manifests.yaml
    - uses: git-commit
      as: commit
      config:
        path: ${{ vars.outPath }}
        messageFromSteps:
        - update-image
    - uses: git-push
      config:
        path: ${{ vars.outPath }}
        branch: ${{ vars.targetBranch }}
    - uses: argocd-update
      config:
        apps:
        - name: kargo-demo-${{ ctx.stage }}
          sources:
          - repoURL: ${{ vars.gitopsRepo }}
            desiredRevision: ${{ outputs.commit.commit }}
```

> __For complete documentation of all of Kargo's built-in promotion steps, refer
> to the [Promotion Steps Reference](../35-references/10-promotion-steps.md).__

Reviewer: ```suggestion
:::info
For complete documentation of all Kargo's built-in promotion steps, refer
to the [Promotion Steps Reference](../35-references/10-promotion-steps.md).
:::
```
